# ã‚­ãƒ¼ãƒ»ã‚³ãƒ³ãƒ•ã‚£ã‚°ã€€ï¼ã€€ã‚¤ãƒ³ãƒ—ãƒƒãƒˆã€€ï¼ã€€ãƒ‘ãƒ¼ã‚µãƒ¼ï¼ˆParserï¼›è§£ææ©Ÿï¼‰
#
#	Godot ã®å…¥åŠ›ã¯æ–‡å­—åˆ—ã§é€ã‚‰ã‚Œã¦ãã‚‹
extends Node


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãƒ¡ãƒ¢ãƒªé–¢é€£
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

# æ–‡å­—åˆ—è§£æç”¨
var re_button = RegEx.new()
var re_lever = RegEx.new()


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# èµ·å‹•æ™‚è¨­å®š
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

func _ready():
	# å…¥åŠ›ã‚¤ãƒ™ãƒ³ãƒˆãŒè¿”ã™æ–‡å­—åˆ—ã€‚ä»•æ§˜ã•ã£ã±ã‚Šåˆ†ã‹ã‚‰ã‚“
	# æœ€å¾Œã«åŠè§’ã‚¹ãƒšãƒ¼ã‚¹ã‚’å…¥ã‚Œã‚‹ã“ã¨ã€‚ `Button 1` ã¨ `Button 10` ã‚’åŒºåˆ¥ã™ã‚‹ãŸã‚ã«
	re_button.compile("Joypad Button (\\d+) ")
	# ä¾‹ï¼š Joypad Motion on Axis 4 (Joystick 2 X-Axis, Left Trigger, Sony L2, Xbox LT) with Value 0.00
	re_lever.compile("Joypad Motion on Axis (\\d+) \\(.*\\) with Value (-?\\d+(?:\\.\\d+)?)")


# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼
# ãã®ä»–
# ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼ãƒ¼

# ä»¥ä¸‹ã®ã‚·ãƒ³ãƒœãƒ«ã‚’è¿”ã™
#
# -1ï¼šã€€è©²å½“ç„¡ã—
# 1000 æœªæº€ã®æ•´æ•°ï¼šã€€ãƒœã‚¿ãƒ³ç•ªå·
# 1000 ä»¥ä¸Šã®æ•´æ•°ï¼šã€€ãƒ¬ãƒãƒ¼ç•ªå·ã«1000ã‚’è¶³ã—ãŸã‚‚ã®
# æ–‡å­—åˆ—ï¼šã€€ã‚­ãƒ¼ãƒœãƒ¼ãƒ‰ã®ã‚­ãƒ¼å
func get_button_symbol_by_text(event_as_text):
	# ğŸ“–ã€€[enum JoyButton:](https://docs.godotengine.org/en/stable/classes/class_%40globalscope.html#enum-globalscope-joybutton)
	# ãƒ¬ãƒãƒ¼ã¯ -1 ï½ 10ã€ ãƒœã‚¿ãƒ³ã¯ -1 ï½ 128 ã¾ã§ã‚ã‚‹ãã†ã 
	var matched = self.re_button.search(event_as_text)
	if matched:
		return int(matched.get_string(1))

	matched = self.re_lever.search(event_as_text)
	if matched:
		return int(matched.get_string(1)) + 1000
	
	print("ï¼»å…¥åŠ›å¤‰æ›ï¼½ã€€event_as_textï¼š" + event_as_text + " type:" + str(typeof(event_as_text)))
	
	if typeof(event_as_text) == TYPE_STRING:
		# ã‚­ãƒ¼ãƒœãƒ¼ãƒ‰ã®ã‚­ãƒ¼ã®åå‰
		return event_as_text
	
	return -1


# å†…éƒ¨ã§ã¯ä½¿ã£ã¦ãªã„ã€‚å¤–éƒ¨å‘ã‘ã€‚
# ãƒ¬ãƒãƒ¼ã®ã‚¤ãƒ™ãƒ³ãƒˆæ–‡å­—åˆ—ã‹ã‚‰ã€-1.0 ï½ 1.0 ã®å€¤ã‚’å–å¾—
# ãƒ¬ãƒãƒ¼ã§ãªã‘ã‚Œã° 0.0 ã‚’è¿”ã™
func get_lever_value_by_text(event_as_text):
	var matched = self.re_lever.search(event_as_text)
	if matched:
		return float(matched.get_string(2))

	return 0.0
